File: Iteration2/CSE3063F24P1_GRP14/src/CourseRegistrationSystem.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CourseRegistrationSystem {
    private Student student;
    //private List<Student> students;
    private List<Course> courses;
    private JSONMethods jsonMethods = new JSONMethods(); // Assuming JSONMethods class handles JSON operations

    // Constructor
    public CourseRegistrationSystem(Student student, List<Course> courses) {
        this.student = student;
        //this.students = students;
        this.courses = courses;
    }


    // Method to add a student to a course
    public void addToEnrollList(Course course, Student student) throws IOException {
        // Check if the student is already enrolled in the course by comparing course IDs
        boolean isAlreadyEnrolled = false;
        for (Course enrolledCourse : student.getEnrolledCourses()) {
            if (enrolledCourse.getCourseId().equals(course.getCourseId())) {
                isAlreadyEnrolled = true;
                break;
            }
        }

        if (isAlreadyEnrolled) {
            System.out.println("Student is already enrolled in this course.");
            return;
        }


        // Enroll the student in the course
        student.getEnrolledCourses().add(course);

        // Update the course's enrollment capacity
        // section.setEnrollmentCapacity(section.getEnrollmentCapacity() - 1);

        // Save the updated student object to JSON file
        jsonMethods.saveStudentToFile(student);

        System.out.println("Student enrolled in course successfully.");
    }

    //Remove the course from the students request list
    public boolean removeCourseFromRequestList(Student student, Course course) throws IOException {
        if(student.getRequestedCourses().size() > 0 && student.getRequestedCourses().contains(course)){
            student.getRequestedCourses().remove(course);
            jsonMethods.saveStudentToFile(student);
            return true; //succesfully removed

        }
        else{
            System.out.println("You did not take any course. Firstly, please request at least one course.");
            return false; //removal failed
        }
    }

    public void removeCourseFromEnrolledList(Student student, Course course) throws IOException {
        if(student.getEnrolledCourses().size() > 0 && student.getEnrolledCourses().contains(course)){
            student.getEnrolledCourses().remove(course);
            jsonMethods.saveStudentToFile(student);

        }
        else{
            System.out.println("You are not enrolled to any course. Firstly, please request at least one course.");
        }
    }

    // Method to list available course sections for a given student
    public List<Course> listAvailableCourses() {
        List<Course> availableCourses = new ArrayList<>();
        List<Course> takenCourses = new ArrayList<>();
        List<String> failedCourses = new ArrayList<>();
        List<String> passedCourses = new ArrayList<>();

        // Geçerli notlar
        List<String> passingGrades = Arrays.asList("AA", "BA", "BB", "CB", "CC");

        // Transcript'teki dersleri kategorize et
        for (Grade grade : student.getTranscript().allGrades()) {
            String gradeValue = grade.getGradeValue();
            Course course = grade.getCourse();

            // Başarısız dersler
            if (gradeValue.equals("FF") || gradeValue.equals("FD")) {
                failedCourses.add(course.getCourseId());
            }
            // Başarılı dersler
            else if (passingGrades.contains(gradeValue)) {
                passedCourses.add(course.getCourseId());
            }
            // Düşük notlar
            else if (gradeValue.equals("DD") || gradeValue.equals("DC")) {
                takenCourses.add(course);
            }
        }

        // Diğer dersleri kontrol et
        for (Course course : courses) {
            // Halihazırda alınan, başarıyla tamamlanan veya requested/enrolled listesinde olan dersleri atla
            if (passedCourses.contains(course.getCourseId()) ||
                    takenCourses.contains(course) ||
                    student.getRequestedCourses().stream().anyMatch(c -> c.getCourseId().equals(course.getCourseId())) ||
                    student.getEnrolledCourses().stream().anyMatch(c -> c.getCourseId().equals(course.getCourseId()))) {
                continue;
            }

            // Prerequisite kontrolü
            if (course.hasPrerequisite()) {
                String prerequisiteId = course.getPrerequisiteLessonId();
                if (!passedCourses.contains(prerequisiteId)) {
                    continue; // Prerequisite geçilmemiş, eklemeyi atla
                }
            }

            // Tüm kontrolleri geçen dersleri listeye ekle
            availableCourses.add(course);
        }

        return availableCourses;
    }







    public String availableCoursesToString(List<Course> availableCourses) {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("%-10s %-40s\n", "Course ID", "Course Name"));
        sb.append("------------------------------------------------------------\n");

        for (Course course : availableCourses) {
            sb.append(String.format("%-10s %-40s\n",
                    course.getCourseId(),
                    course.getCourseName()));
        }

        return sb.toString();
    }



    public void requestInCourse(Course course, Student student) throws IOException {
        // Öğrencinin kayıtlı kurslarını kontrol et
        if (student.getEnrolledCourses().contains(course)) {
            System.out.println("You are already enrolled in this course.");
            return;
        }

        // Öğrencinin daha önce talep ettiği kursları kontrol et
        if (student.getRequestedCourses().contains(course)) {
            System.out.println("You have already requested this course.");
            return;
        }

        System.out.println( " counted :" +countRequestedStudents(jsonMethods.loadAllStudents(), course));
        // Kursun kapasitesini kontrol et
        if (countRequestedStudents(jsonMethods.loadAllStudents(), course) >= course.getEnrollmentCapacity()) {
            System.out.println("hopp waitlistte giriyosun");
            addToWaitList(student, course);
            System.out.println("waitlisttesin");
            jsonMethods.updateCourseInJson(course);
            jsonMethods.updateStudentInJson(student);
            System.out.println("This course is full and cannot accept more students. You are added to wait list of course "
                    + course.getCourseId() + ".");
            return;
        }

        // Öğrencinin requestedCourses listesine ekle
        student.getRequestedCourses().add(course);

        // JSON'da güncelleme
        jsonMethods.updateStudentInJson(student);
        System.out.println( " updated counted :" +countRequestedStudents(jsonMethods.loadAllStudents(), course));

        System.out.println("Successfully requested the course: " + course.getCourseName());
    }
    private void addToWaitList(Student student, Course course){
        course.getWaitList().add(student.getStudentID());
    }

    private int countRequestedStudents(List<Student> allStudents, Course course) {
        int count = 0;
        for (Student student : allStudents) {
            for (Course requestedCourse : student.getRequestedCourses()) {
                if (requestedCourse.getCourseId().equals(course.getCourseId())) {
                    count++;
                    break;
                }
            }
        }
        return count;
    }


}

File: Iteration 3/CourseRegistrationSystem.py
from typing import List
from Student import Student
from Course import Course
from JsonMethods import JsonMethods

class CourseRegistrationSystem:
    def __init__(self, json_methods, student=None, courses=None):
        """
        Initializes the CourseRegistrationSystem.

        Args:
            json_methods (JsonMethods): The instance of JsonMethods.
            student (Student, optional): The student using the system. Defaults to None.
            courses (list, optional): List of courses. Defaults to None.
        """
        self.json_methods = json_methods
        self._student = student  # Allow None for advisor operations
        self._courses = courses if courses is not None else []


    def add_to_enroll_list(self, course: Course, student: Student):
        """
        Enrolls the student in the course if not already enrolled.
        """
        # Reload student data
        student = self.json_methods.load_student(student.get_studentID())

        is_already_enrolled = any(enrolled_course.get_course_id() == course.get_course_id()
                                  for enrolled_course in student.get_enrolled_courses())

        if is_already_enrolled:
            print("Student is already enrolled in this course.")
            return

        # Add course to the student's enrolled courses
        student.get_enrolled_courses().append(course)

        # Save student data to JSON
        self.json_methods.save_student_to_file(student)

        print("Student enrolled in course successfully.")

    def remove_course_from_request_list(self, student: Student, course: Course) -> bool:
        """
        Removes the course from the student's requested courses.
        """
        # Reload student data
        student = self.json_methods.load_student(student.get_student_id())

        requested_courses = student.get_requested_courses()
        if course in requested_courses:
            requested_courses.remove(course)
            self.json_methods.save_student_to_file(student)  # Save updated data
            print("Course removed from requested list.")
            return True
        else:
            print("Course is not in the request list.")
            return False

    def remove_course_from_enrolled_list(self, student: Student, course: Course):
        """
        Removes the course from the student's enrolled courses.
        """
        # Reload student data
        student = self.json_methods.load_student(student.get_student_id())

        enrolled_courses = student.get_enrolled_courses()
        if course in enrolled_courses:
            enrolled_courses.remove(course)
            self.json_methods.save_student_to_file(student)  # Save updated data
            print("Course removed from enrolled list.")
        else:
            print("Course is not in the enrolled list.")

    def list_available_courses(self, logged_in_student: Student) -> List[Course]:
        """
        Returns a list of courses available for the student to enroll in.
        """
        # Reload student data
        logged_in_student = self.json_methods.load_student(logged_in_student.get_studentID())

        available_courses = []
        taken_courses = []
        failed_courses = []
        passed_courses = []
        passing_grades = {"AA", "BA", "BB", "CB", "CC"}

        # Ensure transcript is a proper object
        transcript = logged_in_student.get_transcript()
        if isinstance(transcript, dict):
            from Transcript import Transcript
            transcript = Transcript(**transcript)

        # Process grades to categorize courses
        for grade in transcript.all_grades():
            grade_value = grade.get_grade_value()
            course = grade.get_course()

            if grade_value in {"FF", "FD"}:
                failed_courses.append(course.get_course_id())
            elif grade_value in passing_grades:
                passed_courses.append(course.get_course_id())
            elif grade_value in {"DD", "DC"}:
                taken_courses.append(course)

        # List courses that are available to enroll
        for course in self._courses:
            if (course.get_course_id() in passed_courses or
                    course in taken_courses or
                    any(c.get_course_id() == course.get_course_id() for c in
                        logged_in_student.get_requested_courses()) or
                    any(c.get_course_id() == course.get_course_id() for c in logged_in_student.get_enrolled_courses())):
                continue

            # Check prerequisites
            if course.get_prerequisite() and course.get_prerequisite_lesson_id() not in passed_courses:
                continue

            available_courses.append(course)

        return available_courses

    

    def request_in_course(self, course: Course, student: Student):
        """
        Adds a course to the student's requested courses if eligible.
        """
        # Reload student data
        student = self.json_methods.load_student(student.get_studentID())

        # Capacity check
        if course.get_current_capacity() >= course.get_enrollment_capacity():
            print("This course is full and cannot accept more students.")
            return

        # Check if the course has already been requested
        if course in student.get_requested_courses():
            print("You have already requested this course.")
            return

        # Add the course to the student's requested courses list
        student.get_requested_courses().append(course)

        # Save updated student data
        self.json_methods.save_student_to_file(student)

        # Update the `_student` attribute to reflect the changes
        self._student = self.json_methods.load_student(student.get_studentID())

        print(f"Successfully requested the course: {course.get_course_name()}")


    
